### DISCUSSION TITLE
What's the difference between includePaths and fileMatch? Why is that sometimes a filename needs to be added in both places to be processed?

### DISCUSSION THREAD

**Discussion by phaninra** [renovatebot/renovate#18524] [Request Help] [answered]
The objective is to process only the selected package files and ignore everything else. 

The relevants parts from the repo file tree
```
root
- gradle/sample.toml // .toml in this dir is an officially recognized location, i.e. part of https://docs.renovatebot.com/modules/manager/gradle/#file-matching
- arbitrary/nested/sample.toml // this is not since .toml files are not recognized outside of gradle dir.
```

I've experimented with a few combinations

1) `enabledManagers: ["gradle"]` : 
    * `gradle/sample.toml` and many other files in the repo were upadted. `arbitrary/nested/sample.toml` was not updated. All of that just as expected.

2) `gradle: { includePaths: ["**/sample\\.toml"]}` : 
    * Only `gradle/sample.toml` was updated. 

3) `gradle: { includePaths: ["**/sample\\.toml"], fileMatch: ["sample.toml"]}`: 
    * Both `sample.toml` files were updated and nothing else. 

4) `gradle: { fileMatch: ["sample.toml"]}`: 
    * Everything in the repo got updated including the `arbitrary/nested/sample.toml`.

Why did the (2) `"**/sample\\.toml"` process only the `gradle/sample.toml` but not the `arbitrary/nested/sample.toml `?

Why is that the `sample.toml` needs to be added to both the `fileMatch` and `includePaths` for the `arbitrary/nested/sample.toml` to be recognized?  (Ref (2) and (3)) 

What's the difference between `includePaths` and `fileMatch`? As a side note, maybe I'm missing something but it's confusing that they follow different matching algos and that's not mentioned in the `includePaths` docs.

**Comment by Churro** [comment, marked as answer]
- `includePaths` limits the set of files renvate processes with any manager
  - E.g., a subfolder of your repository containing Gradle build files + Kubernetes manifests
- `fileMatch` is set for a specific manager (e.g. gradle) and limits the files processed by that specific manager
  - E.g., version catalogs in `*.toml` files only

So, for `fileMatch` to capture your files, they need to be matched by `includePaths` first. This means:

3. `gradle: { includePaths: ["**/sample\\.toml"], fileMatch: ["sample.toml"]}` -> both conditions satisified
4. `gradle: { fileMatch: ["sample.toml"]}` -> no `includePaths` specified means pass all files to managers. With Gradle manager, `fileMatch` is then the only constraint to decide whether this is procesed

On the syntax, `includePaths` follows the [minimatch](https://github.com/isaacs/minimatch) format, `fileMatch` uses pure RegEx.
